clients/drcachesim/tracer/tracer.cpp

```diff
@@ -1330,6 +1330,27 @@ event_pre_syscall(void *drcontext, int sysnum)
         return true;
     if (BUF_PTR(data->seg_base) == NULL)
         return true; /* This thread was filtered out. */
+
+    // If we just switched modes in this thread in a block ending in a syscall,
```

Comment:
Maybe I'm missing something: event_inscount_app_instruction seems to instrument a hit_instr_count_threshold call at each instruction (I see that it is called only when the inlined-counter reaches the instr-counting-threshold). So tracing_mode would flip to BBDUP_MODE_TRACE as soon as we hit the instr-counting threashold. But what if we flip it only at the end of the block, after the syscall callbacks if any. This way there'll be parity between other actions being performed while in that block (e.g. this event_pre_syscall checking tracing_mode). Or are there complexities wrt not knowing whether the syscall will return?

Reply:
Syscall callbacks are after the block so waiting until the end of the block would not help.